### PR TITLE
adding a loading status

### LINKED ISSUE #227: (refactor/bug) Trial Balance needs a loading indicator
The Trial Balance freezes for a variable amount of time depending on the number of records there are in the Posting Journal before displaying information.  It looks like this:

trialbalancefreezing
_Fig 1: Trial Balance freezing during loading_

This is not ideal as the user isn't sure if the application is broken or not.  If we put a loading indicator on the Trail Balance grid (maybe even with the `bhGridLoading` indicator), this will help the user feel like something is happening.

### PR DESCRIPTION
**Pull request by DedrickEnc**
This PR closes #227, it is showing the loading indicator during data loading like this : 
loading trialbalance

Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

File: client/src/partials/journal/modals/trialBalanceMain.body.html [hunk 1]
```diff
@@ -2,6 +2,7 @@
   <div class="row">
     <div class="col-xs-4 pull-left">
       <button
+        ng-if="!TrialBalanceMainBodyCtrl.loading"
```
Comment (jniles) [inline client/src/partials/journal/modals/trialBalanceMain.body.html]: You already have a `ng-show`.  Could you just combine all these?  For example:
```js
var hasErrorCondition = vm.feedback.hasError || vm.feedback.hasWarning;
vm.showErrorButton =  !vm.loading && hasErrorCondition;
```

File: client/src/partials/journal/modals/trialBalanceMain.body.html [hunk 2]
```diff
@@ -29,11 +31,21 @@
...
+    >
+    </bh-grid-loading-indicator>
+  </div>
   <div
+    ng-if="!TrialBalanceMainBodyCtrl.loading"
     id="main-grid"
     ui-grid="TrialBalanceMainBodyCtrl.gridOptions"
     ui-grid-auto-resize
     ui-grid-grouping>
+    </bh-grid-loading-indicator>
```
Comment (jniles) [inline client/src/partials/journal/modals/trialBalanceMain.body.html]: Hmm... this is broken HTML.

Normally the grid loading indicator goes entirely inside the grid.
```html
<div id="main-grid" ...>
    <bh-grid-loading-indicator
      loading-state="TrialBalanceMainBodyCtrl.loading"
      empty-state="TrialBalanceMainBodyCtrl.gridOptions.data.length === 0"
      error-state="TrialBalanceMainBodyCtrl.hasError">
    </bh-grid-loading-indicator>
</div>
```

I also think you could give the grid a default size, that way the modal doesn't grow and shrink a lot with state changes.  It makes for a better UX.

Comment (DedrickEnc) [inline client/src/partials/journal/modals/trialBalanceMain.body.html]: Inside the grid, the grid column will not be colored appropriately.

Comment (jniles) [inline client/src/partials/journal/modals/trialBalanceMain.body.html]: What color should the column be while loading?  Can you show a screenshot to see?

**Comment by DedrickEnc** [comment]
When error in trial balance, column color must be red so putting ```bhloadingIndicator``` inside the grid give me this : 

color error

**Comment by jniles** [comment]
@DedrickEnc , looks like a CSS selector issue.  What is the code to color the columns red?

**Comment by DedrickEnc** [comment]
In the controller, here is the code : 

```
  trialBalanceService.checkTransactions(vm.dataByTrans)
    .then(function(error) {
      var cssClass = null;
      errorList = error.data;
      vm.feedBack = trialBalanceService.getFeedBack(errorList); //getting a feedback object to customize the grid
      vm.isInvalid = vm.feedBack.hasError || vm.feedBack.hasWarning;
      cssClass = trialBalanceService.getCSSClass(vm.feedBack);
      $state.current.data.checkingData = {errors : errorList, feedBack : vm.feedBack, cssClass : cssClass};
      $state.current.data.checked = true;

      columns.forEach(function (col) {
        col.headerCellClass = cssClass;
      });
    })
    .catch(errorHandler)
    .finally(function () {
      vm.loading = false;
      vm.showErrorButton = vm.isInvalid && !vm.loading;
    });
```

**Comment by jniles** [comment]
@DedrickEnc, this has nothing to do with the `bhGridLoadingComponent` is behaving as expected.

The problem is that you are using an `ngIf` on the grid.  The order in the code that works is this (pseudocode)
```js
.then(() => {

  // change the css class
  cssClass = trialBalanceService.getCSSClass(vm.feedBack);

  // change grid options
  columns.forEach(function (col) {
     col.headerCellClass = cssClass;
  });

  // ui-grid has no idea anything changed!
})
.catch(errorHandler)
.finally(function () {

   // this will run an angular $digest() loop because it toggles an ngIf in the HTML
   // therefore, the entire page/grid will redraw and update it's colors
  vm.loading = false;
});
```

Since you are no longer using an `ngIf`, it is not necessary to redraw the entire page to update the grid color.  Instead, you can just tell the grid the columns changed.

```js
.then(() => {

  // change the css class
  cssClass = trialBalanceService.getCSSClass(vm.feedBack);

  // change grid options
  columns.forEach(function (col) {
     col.headerCellClass = cssClass;
  });

  // ui-grid has no idea anything changed!  Let's tell it.
  gridApi.core.notifyDataChange(uiGridConstants.dataChange.COLUMN);
}) 
.catch(errorHandler)
.finally(function () {

  // there is no ngIf, so this no longer triggers a redraw of the grid.  Yay!
  vm.loading = false;
});
```

Remember that whenever you update ui-grid's properties, it should be notified by emitting an event.

**Comment by jniles** [comment]
Updated with better explanations, I typed it up in a hurry...

**Comment by jniles** [comment]
All tests pass!  Merging...

**Pull request merged**